man/man8/genl.8
.TH GENL 8 "29 Oct 2015" "iproute2" "Linux"
.SH NAME
genl \- generic netlink utility frontend
.SH SYNOPSIS
.in +8
.ti -8
.BR genl "[ " OPTIONS " ] "  OBJECT
.ti -8
.BR genl " { " -V [ ersion "] | " -h [ elp "] }"

.ti -8
.IR OPTIONS " := { "
\fB\-s\fR[\fItatistics\fR] |
\fB\-d\fR[\fIetails\fR] |
\fB\-j\fR[son\fR] |
\fB\-p\fR[retty\fR] }

.ti -8
.IR OBJECT " := { "
.B ctrl
.IR CTRL_OPTS " }"

.ti -8
.IR CTRL_OPTS " := { "
.BR help " | " list " | " monitor " | " get
.IR PARMS " }"

.ti -8
.IR PARMS " := { "
.B name
.IR NAME " | "
.B id
.IR ID " }"
.SH DESCRIPTION
The
.B genl
utility provides a simple frontend to the generic netlink library. Although it's
designed to support multiple
.IR OBJECT s,
for now only the
.B ctrl
object is available, which is used to query the generic netlink controller.
.SS ctrl
The generic netlink controller can be queried in various ways:
.TP
.B help
This command just prints a help text for the
.B ctrl
object.
.TP
.B list
Show the registered netlink users.
.TP
.B monitor
Listen for generic netlink notifications.
.TP
.B get
Query the controller for a given user, identified either by
.BR name " or " id .
.SH OPTIONS
genl supports the following options.
.TP
.B \-h, \-help
Show summary of options.
.TP
.B \-V, \-Version
Show version of program.
.TP
.B \-s, \-stats, \-statistics
Show object statistics.
.TP
.B \-d, \-details
Show object details.
.TP
.B -j, \-json
Output results in JavaScript Object Notation (JSON).
.TP
.B -p, \-pretty
Add indentation for readability.
.TP
.B \-r, \-raw
Dump raw output only.
.SH SEE ALSO
.BR ip (8)
.br
.SH AUTHOR
genl was written by Jamal Hadi Salim <hadi@cyberus.ca>.
.PP
This manual page was written by Petr Sabata <contyk@redhat.com>.


ChangeLog
2006-03-21  James Lentini <jlentini@netapp.com>
	
	* Increase size of hw address allowed for ip neigh to allow
	  for IB.

2006-03-14  Russell Stuart <russell-lartc@stuart.id.au>
	
	* Fix missing memset in tc sample
	* Fixes for tc hash samples
	* Add sample divisor

2006-03-10  Alpt <alpt@freaknet.org>

	* Add more rt_proto values

2006-03-10 Dale Sedivec <darkness@caliginous.net>
	
	* Warn when using "handle" instead of "classid" with "tc class"

2006-03-10  Jean Tourrilhes <jt@hpl.hp.com>

	* Fix endless loop in netlink error handling


2006-03-10  Stephen Hemminger  <shemminger@osdl.org>

	* Change default lnstat count to 1
	* Update to 2.6.16 headers
	* Add fake version of include/linux/socket.h to fix warnings

2006-01-12  Patrick McHardy <kaber@trash.net>

	* Handle DCCP in ipxfrm.c to allow using port numbers in the selector.

2006-01-10  Masahide NAKAMURA <nakam@linux-ipv6.org>

	* Add ip link ntable

2006-01-10  Stephen Hemminger  <shemminger@osdl.org>

	* Update headers to santized kernel 2.6.15
	* Fix ipv6 priority option in u32 
	
2006-01-03 Alpt <alpt@freaknet.org>

	* Ip man page addition

2006-01-03  Jamal Hadi Salim <hadi@znyx.com>

	* Documentation for ifb

2005-12-09  Stephen Hemminger  <shemminger@osdl.org>

	* Add corrupt feature to netem

2005-12-02  Stephen Hemminger  <shemminger@osdl.org>

	* Backout ambigious ip command matches

2005-11-22  Stephen Hemminger  <shemminger@osdl.org>

	* Handle ambigious ip command matches

2005-11-22  Patrick McHardy <kaber@trash.net>

	* Add back ip command aliases

2005-11-07  Masahide NAKAMURA <nakam@linux-ipv6.org>

	* Updating for 2.6.14
	- Show UPD{SA,POLICY} message information from kernel instead of error
	- Add lengh check of deleting message from kernel
	- Use macro for struct xfrm_user{sa,policy}_id

	* Minor fix:
	- Add fflush at the end of normal dump

2005-11-01  Jamal Hadi Salim <hadi@znyx.com>

	* Fix handling of XFRM monitor and state

2005-11-01  Stephen Hemminger  <shemminger@osdl.org

	* Update to 2.6.14 sanitized headers

2005-10-24  Patrick McHardy <kaber@trash.net>

	* Fix ip commnad shortcuts

2005-10-12  Stephen Hemminger  <shemminger@osdl.org>

	* Add more CBQ examples from Fedora Core
	* Fix buffer overrun in iproute because of bits vs. bytes confusion
	
2005-10-12  Jamal Hadi Salim <hadi@znyx.com>

	* Fix ip rule flush, need to reopen rtnl

2005-10-07  Stephen Hemminger  <shemminger@osdl.org>

	* Reenable ip mroute

2005-10-07  Mike Frysinger <vapier@gentoo.org>

	* Handle pfifo_fast that has no qopt without segfaulting

2005-10-05  Mads Martin Joergensen <mmj@suse.de>

	* Trivial netem ccopts

2005-10-04  Jerome Borsboom <j.borsboom@erasmusmc.nl>

	* Fix regression in ip addr (libnetlink) handling

2005-09-21  Stephen Hemminger  <shemminger@osdl.org>

	* Fix uninitialized memory and leaks with valgrind
	  Reported by Redhat

2005-09-01   Mike Frysinger <vapier@gentoo.org>

	* Fix build issues with netem tables (parallel make and HOSTCC)
	
2005-09-01  Stephen Hemminger  <shemminger@osdl.org>

	* Integrate support for DCCP into 'ss' (from acme)
	* Add -batch option to ip.
	* Update to 2.6.14 headers

2005-09-01  Eric Dumazet <dada1@cosmosbay.com>
	
	* Fix lnstat : First column should not be summed

2005-08-16  Stephen Hemminger  <shemminger@osdl.org>

	* Limit ip route flush to 10 rounds.
	* Cleanup ip rule flush error message
	
2005-08-08  Stephen Hemminger  <shemminger@osdl.org>

	* Update to 2.6.13+ kernel headers
	* Fix array overrun in paretonormal
	* Fix ematch to not include dropped fields from skb.
	
2005-07-14  Thomas Graf <tgraf@suug.ch>

	* Make ematch bison/lex build with common flex
	
2005-07-10  Stephen Hemminger  <shemminger@osdl.org>
	
	* Fix Gcc 4.0 build warnings signed/unsigned

2005-06-23  Jamal Hadi Salim <hadi@znyx.com>

	* Fix for options process with ipt

2005-06-23  Thomas Graf <tgraf@suug.ch>
	
	* Add extended matches (nbyte, cmp, u32, meta)
	* Add basic classifier
	* Fix clean/distclean makefile targets
	* update local header file copies
	* IPv4 multipath algorithm selection support
	* cscope Makefile target
	* Fix off-by-one while generating argument vector
	  in batched mode.
	* Assume stdin if no argument is given to -batch

2005-06-22  Stephen Hemminger  <shemminger@osdl.org>

	* Update include files to 2.6.12
	* Add ss support for TCP_CONG

2005-06-13  Steven Whitehouse <steve@chygwyn.com>

	* Decnet doc's update

2005-06-07  Stephen Hemminger  <shemminger@osdl.org>

	* Fix 'ip link' map to handle case where device gets autoloaded
	  by using if_nametoindex as fallback
	* Device indices are unsigned not int.

2005-06-07   Masahide NAKAMURA <nakam@linux-ipv6.org>
	
	* [ip] show timestamp when using '-t' option.
	* [ip] remove duplicated code for expired message of xfrm.
	* [ip] add "deleteall" command for xfrm;
	  "flush" uses kernel's flush interface and
	  "deleteall" uses legacy iproute2's flush feature like
	   getting-and-deleting-for-each.

2005-03-30  Stephen Hemminger  <shemminger@osdl.org>

	* include/linux/netfilter_ipv4/ip_tables.h dont include compiler.h
	  because it isn't needed and not on all systems
	* Update rtnetlink.h and pkt_cls.h to be stripped versions
	  of headers from 2.6.12-rc1

2005-03-30  Jamal Hadi Salim <hadi@znyx.com>

	* Proper verison of iptables headers (from 1.3.1)
	* Set revision file in m_ipt
	* Fix action_util naming in mirred
	* don't call ll_init_map in mirred

2005-03-19  Thomas Graf <tgraf@suug.ch>

	* Warn about wildcard deletions and provide IFA_ADDRESS upon
	  deletions to enforce prefix length validation for IPv4.
	* Fix netlink message alignment when the last routing attribute added
	  has a data length not aligned to RTA_ALIGNTO.
	
2005-03-30  Masahide NAKAMURA <nakam@linux-ipv6.org>
	
	* ipv6 xfrm allocspi and monitor support.
	
2005-03-29  Stephen Hemminger  <shemminger@osdl.org>

	* switch to stack for netem tables

2005-03-18  Stephen Hemminger  <shemminger@osdl.org>

	* add -force option to batch mode

genl/genl.c
/* SPDX-License-Identifier: GPL-2.0-or-later */
/*
 * genl.c		"genl" utility frontend.
 *
 * Authors:	Jamal Hadi Salim
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>
#include <dlfcn.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <string.h>
#include <errno.h>
#include <linux/netlink.h>
#include <linux/rtnetlink.h> /* until we put our own header */
#include "version.h"
#include "utils.h"
#include "genl_utils.h"

int show_stats;
int show_details;
int show_raw;
int json;

static void *BODY;
static struct genl_util *genl_list;


static int print_nofopt(struct nlmsghdr *n, void *arg)
{
	fprintf((FILE *) arg, "unknown genl type ..\n");
	return 0;
}

static int parse_nofopt(struct genl_util *f, int argc, char **argv)
{
	if (argc) {
		fprintf(stderr,
			"Unknown genl \"%s\", hence option \"%s\" is unparsable\n",
			f->name, *argv);
		return -1;
	}

	return 0;
}

static struct genl_util *get_genl_kind(const char *str)
{
	void *dlh;
	char buf[256];
	struct genl_util *f;

	for (f = genl_list; f; f = f->next)
		if (strcmp(f->name, str) == 0)
			return f;

	snprintf(buf, sizeof(buf), "%s.so", str);
	dlh = dlopen(buf, RTLD_LAZY);
	if (dlh == NULL) {
		dlh = BODY;
		if (dlh == NULL) {
			dlh = BODY = dlopen(NULL, RTLD_LAZY);
			if (dlh == NULL)
				goto noexist;
		}
	}

	snprintf(buf, sizeof(buf), "%s_genl_util", str);

	f = dlsym(dlh, buf);
	if (f == NULL)
		goto noexist;
reg:
	f->next = genl_list;
	genl_list = f;
	return f;

noexist:
	f = calloc(1, sizeof(*f));
	if (f) {
		strncpy(f->name, str, 15);
		f->parse_genlopt = parse_nofopt;
		f->print_genlopt = print_nofopt;
		goto reg;
	}
	return f;
}

static void usage(void) __attribute__((noreturn));

static void usage(void)
{
	fprintf(stderr,
		"Usage: genl [ OPTIONS ] OBJECT [help] }\n"
		"where  OBJECT := { ctrl etc }\n"
		"       OPTIONS := { -V[ersion] | -s[tatistics] | -d[etails] | -r[aw] |\n"
		"                    -j[son] | -p[retty] }\n");
	exit(-1);
}

int main(int argc, char **argv)
{
	while (argc > 1) {
		if (argv[1][0] != '-')
			break;
		if (matches(argv[1], "-stats") == 0 ||
		    matches(argv[1], "-statistics") == 0) {
			++show_stats;
		} else if (matches(argv[1], "-details") == 0) {
			++show_details;
		} else if (matches(argv[1], "-raw") == 0) {
			++show_raw;
		} else if (matches(argv[1], "-Version") == 0) {
			printf("genl utility, iproute2-%s\n", version);
			exit(0);
		} else if (matches(argv[1], "-json") == 0) {
			++json;
		} else if (matches(argv[1], "-pretty") == 0) {
			++pretty;
		} else if (matches(argv[1], "-help") == 0) {
			usage();
		} else {
			fprintf(stderr,
				"Option \"%s\" is unknown, try \"genl -help\".\n",
				argv[1]);
			exit(-1);
		}
		argc--;	argv++;
	}

	if (argc > 1) {
		struct genl_util *a;
		int ret;

		a = get_genl_kind(argv[1]);
		if (!a) {
			fprintf(stderr, "bad genl %s\n", argv[1]);
			exit(-1);
		}

		ret = a->parse_genlopt(a, argc-1, argv+1);
		return ret;
	}

	usage();
}
